astropy/io/fits/tests/test_image.py

```diff
@@ -403,6 +404,7 @@ def test_verification_on_output(self):
     def test_section(self):
         # section testing
         fs = fits.open(self.data("arange.fits"))
+        assert fs[0].section.dtype == np.dtype(BITPIX2DTYPE[fs[0]._bitpix])
```

Comment:
Seems better to hardcode the expected type here instead of relying on BITPIX2DTYPE (which is not necessarily the correct type):
```suggestion
        assert fs[0].section.dtype == np.int32
```

Also can you please add some tests with files that include scaling, e.g. `fixed-1890.fits` and `scale.fits` ?

Reply:
This is actually returning `np.float64` as the `dtype`, though looking at `_dtype_for_bitpix`, I think this might be expected?

Reply:
hmm actually no.. I forgot that `_dtype_for_bitpix` only gives the correct dtype for hdus where the data must be scaled, and annoyingly we don't have an easy way to get the final dtype :(
I think what we need is
```
if  not hdu._do_not_scale_image_data and (hdu._orig_bscale != 1 or hdu._orig_bzero != 0)
    return np.dtype(BITPIX2DTYPE[self.hdu._bitpix])
else:
    return self.hdu._dtype_for_bitpix()
```
which is used in several places in the code. Ideally we should expose that as a property but since this is a bugfix PR let's keep this addition for later.

Reply:
Btw, `fitsinfo` shows the dtype after rescaling (with the same method). And this is also why I suggested to test different files:
arange.fits returns int32
fixed-1890.fits rescales to uint16
scale.fits rescales to float32

Reply:
It's really good to know about `fitsinfo`, thank you! I just pushed a new commit with this change.